### DISCUSSION TITLE
Modbus forwarder

### DISCUSSION THREAD

**Discussion by michalcharvat** [veista/nilan#51] [Q&A]
Hi, do you have any experience with modbus forwarder? Your extension works pretty well until I try change for example top boiler temperature. The value is not changed in HA, but also it looks like sometimes there is an error on forwarder and sometimes there is no error. I am using modbus2usb for years but currently I have some time to create forwarder to allow connect HA to Nilan. Whereas the forwarder is so simple I am not sure where should I take a look. Do you have any hint? Thanks

**Comment by veista** [comment]
I'm sorry. I do not understand the question.

For the Integration to work  you need either a rs485 to usb adaptor, FTDI chip based usually work the best, or if wiring is not possible, you can use modbus RTU to ModBus TCP bridge.

**Comment by michalcharvat** [comment]
I am sorry for missunderstanding. I use RS485 to USB adapter however I would like to access modbus via TCP/IP which is possible with forwarder written in python. The simple forwarder I use works pretty well until I have to change any value - for example temperature. Do you have any idea where could be the problem? 

```import argparse
import asyncio
import logging
import signal

from pymodbus.client import ModbusSerialClient
from pymodbus.datastore import ModbusServerContext
from pymodbus.datastore.remote import RemoteSlaveContext
from pymodbus.server.async_io import ModbusTcpServer


FORMAT = "%(asctime)-15s %(levelname)-8s %(module)-15s:%(lineno)-8s %(message)s"
logging.basicConfig(format=FORMAT)
_logger = logging.getLogger()


def raise_graceful_exit(*_args):
    """Enters shutdown mode"""
    _logger.info("receiving shutdown signal now")
    raise SystemExit


class SerialForwarderTCPServer:
    """SerialRTU2TCP Forwarder Server"""

    def __init__(self):
        """Initialize the server"""
        self.server = None

    async def run(self):
        """Run the server"""
        port, server_port, server_ip, slaves = get_commandline()
        client = ModbusSerialClient(method="rtu", port=port, baudrate=19200, bytesize=8, parity="E", stopbits=1, timeout=2, debug=True)
        message = f"RTU bus on {port}"
        _logger.info(message)
        store = {}
        for i in slaves:
            store[i] = RemoteSlaveContext(client, unit=i)
        context = ModbusServerContext(slaves=store, single=False)
        self.server = ModbusTcpServer(
            context, address=(server_ip, server_port), allow_reuse_address=True, debug=True
        )
        message = f"serving on {server_ip} port {server_port}"
        _logger.info(message)
        message = f"listening to slaves {context.slaves()}"
        _logger.info(message)
        await self.server.serve_forever()

    async def stop(self):
        """Stop the server"""
        if self.server:
            await self.server.shutdown()
            _logger.info("TCP server is down")

def get_commandline():
    """Read and validate command line arguments"""
    logchoices = ["critical", "error", "warning", "info", "debug"]

    parser = argparse.ArgumentParser(description="Command line options")
    parser.add_argument("--log", help=",".join(logchoices), default="info", type=str)
    parser.add_argument(
        "--port", help="RTU serial port", default="/dev/ttyUSB0", type=str
    )
    parser.add_argument("--server_port", help="server port", default=502, type=int)
    parser.add_argument("--server_ip", help="server IP", default="127.0.0.1", type=str)
    parser.add_argument(
        "--slaves", help="list of slaves to forward", type=int, nargs="+"
    )

    args = parser.parse_args()

    # set defaults
    _logger.setLevel(
        args.log.upper() if args.log.lower() in logchoices else logging.INFO
    )
    if not args.slaves:
        args.slaves = {30}
    return args.port, args.server_port, args.server_ip, args.slaves


if __name__ == "__main__":
    server = SerialForwarderTCPServer()
    try:
        signal.signal(signal.SIGINT, raise_graceful_exit)
        asyncio.run(server.run())
    finally:
        asyncio.run(server.stop())```

**Comment by veista** [comment]
Can't help you with that. See #13

**Reply by michalcharvat** [reply]
Hmm do not tell me there is so huge difference between RTU and TCP modbus. I am able to do everything via RTU but when I try to do same thing via TCP I am stack. TCP server also uses RTU method. I will take a look into your projects.. How do you write data into registers? All at once? When I try change lets say temperature what registers are set? Only the one with temperature? The strange thing is I am able to write to register 1700 which is normally available when I communicate with unit over minimalmodbus but I am unable to write anything into it when I communicate via TCP forwarder .. strange pretty strange.

**Reply by veista** [reply]
There is a difference between RTU and TCP. You can change the protocol in device.py. type: "rtuovertcp"
